Reexport generated macro

arcnmx opened the pull request:
Fixes #18 by allowing definition of slices from non-crate-root contexts such as a module to work as one would expect. Feels like a kind of hacky solution, but I'm familiar enough with rust's macro path resolution rules to know of a better one. I included a 2015 edition test just to be sure that it doesn't break anything (and I'm somewhat surprised that it works?).

arcnmx commented:
(sorry for all the force-pushes, I kept finding very small cleanups to make, I'm done now)

dtolnay commented on impl/src/derive.rs:
```diff
@@ -7,21 +7,25 @@ use crate::linker;
         #[doc(hidden)]
-        #[macro_export]
-        macro_rules! #ident {
+        #[macro_export(#ident)]
```
What does the ident on this line do?

arcnmx replied:
Hm, the intention was that it be re-exported as `crate::ident!` for compatibility with the current behaviour, mostly because I assumed 2015 would require it. Seems to not though!

arcnmx commented on Cargo.toml:
```diff
@@ -9,6 +9,11 @@ documentation = "https://docs.rs/linkme"
 
+[[test]]
+name = "module_2015"
+src = "tests/module_2015.rs"
```
Oops, yeah, that was meant to be `path`. It works anyway because the test name matches the source filename, so maybe I should just remove it?

dtolnay replied:
If it works anyway, you should remove it.

arcnmx commented:
Addressed things and cleaned it up a bit, but was unable to get it to work with rustc 1.31.x - which didn't like the `use` like because `imports can only refer to extern crate names passed with `--extern` on stable channel (see issue #53130)`. It seems that `use something` to refer to current scope only started working in 1.32.0? I'm not sure of a way to work around this, though there's always the option of using a build script to revert to the previous approach (and just accept that only definitions at the crate root will work with 1.31).

(there were also some weird visibility issues on 1.31.x but I think those have workarounds)

dtolnay commented:
Published in 0.1.6.
